### PR TITLE
cmake: improve dependencies in incremental builds

### LINKED ISSUE #26304: When CUDA is enabled, recompiling after editing aten/src/ATen/core/dispatch/DispatchTable.h doesn't result in all necessary files getting recompiled
Other files may also be susceptible, but the issue reproduces 100% on `aten/src/ATen/core/dispatch/DispatchTable.h`

Steps to reproduce:
1. Build PyTorch `python setup.py develop`
2. Edit the "Didn't find kernel to dispatch to" error message to insert some string. Example diff:

```
diff --git a/aten/src/ATen/core/dispatch/DispatchTable.h b/aten/src/ATen/core/dispatch/DispatchTable.h
index 7b4af315f1..3a5879cbda 100644
--- a/aten/src/ATen/core/dispatch/DispatchTable.h
+++ b/aten/src/ATen/core/dispatch/DispatchTable.h
@@ -263,7 +263,7 @@ private:
       }
 
       const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
-      TORCH_CHECK(false, "Didn't find kernel to dispatch to for operator '", operator_name_,
+      TORCH_CHECK(false, "ARGLEBARGLEFARR Didn't find kernel to dispatch to for operator '", operator_name_,
                "'. Tried to look up kernel for dispatch key '", dispatch_key_str,
                "'. Registered dispatch keys are: ", listAllDispatchKeys());
   }
```

3. Rebuild PyTorch `python setup.py develop`
4. Run `python -c "import torch; torch.cat([], dim=1)`

Expected result: you see ARGLEBARGLEFARR
Actual result: you don't see it

I did an experiment where I did one of these botched recompiles, and then compared the set of files which had ARGLEBARGLEFARR with the set of files that had the normal error message. What I found is that CUDA generated files had not been recompiled, but CPU (and final library) all had the string. The fact that the CUDA files are not recompiling seems problematic; however, I don't understand how this could have caused the problem, as the test I made above is CPU (perhaps some sort of weak symbol situation?)

cc @ezyang @gchanan @zou3519 @smessmer

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #26304

Test procedure:
With ninja:
[x] Build a clean checkout
[x] Build again. Result: Only 10 libraries are (needlessly) linked again, the extra delay on a 24-core machine is <10s.
[x] Build for the third time. Result: Virtually instantaneous, with no extra rebuilding.
[x] Modify DispatchTable.h. Build again. Result: `.cu` files are rebuilt, as well as many `.cpp` files
[x] Build for the fifth time. Result: Virtually instantaneous, with no extra rebuilding.
[x] Touch one of the `.depend` files. Build again. Result: Only 10 libraries are (needlessly) linked again, the extra delay on a 24-core machine is <10s.

Without ninja:
[x] Build a clean checkout
[x] Build again. Result: There is some unnecessary rebuilding. But it was also happening before this change.
[x] Build for the third time. Result: Virtually instantaneous, with no extra rebuilding.

### CODE REVIEW

File: tools/setup_helpers/cmake.py [hunk 1]
```diff
@@ -338,19 +338,3 @@ def build(self, my_env):
...
-        # have a timestamp newer than build.ninja's timestamp.
-        # There is a cmake bug with the Ninja backend, where the .cu.depend files
-        # are still compiling by the time the build.ninja timestamp is updated,
-        # so the .cu.depend file's newer timestamp is screwing with ninja's incremental
-        # build detector.
-        # This line works around that bug by manually updating the build.ninja timestamp
-        # after the entire build is finished.
-        ninja_build_file = os.path.join(self.build_dir, 'build.ninja')
-        if os.path.exists(ninja_build_file):
-            os.utime(ninja_build_file, None)
```
Comment (Baranowski) [inline tools/setup_helpers/cmake.py:356]: Things might have been different when this comment was first made. But now I believe the `.depend` files are generated after the `build.ninja` file by design. Yes, it causes the `build.ninja` file to be regenerated, as well as numerous `.cmake` and `.pre-gen` files. However, this, regeneration takes only few seconds, and does not in itself trigger other building steps.

Touching `build.ninja` suppresses regeneration of build.ninja. As a result. it masks triggers for rebuild. I have so far identified three bogus triggers, patched in this PR. The problem is, touching `build.ninja` also masks legitimate triggers. That's what was causing the issue this PR is attempting to solve.

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -1088,7 +1085,12 @@ if(USE_ROCM)
   else()
     caffe2_update_option(USE_ROCM OFF)
   endif()
+endif()
 
+# ---[ ROCm
+if(USE_ROCM)
+  # We check again for USE_ROCM because it might have been set to OFF
+  # in the if above
```
Comment (Baranowski) [inline cmake/Dependencies.cmake:1093]: Before this change, the first run of `cmake` will add `-isystem /opt/rocm/hip/include (...)` to the compilation command whether ROCM is present or not. It will be removed on the second and subsequent runs, which causes a lot of unnecessary rebuilding.

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -854,9 +854,6 @@ if(pybind11_FOUND)
 else()
     message(STATUS "Using third_party/pybind11.")
     set(pybind11_INCLUDE_DIRS ${CMAKE_CURRENT_LIST_DIR}/../third_party/pybind11/include)
-    install(DIRECTORY ${pybind11_INCLUDE_DIRS}
-            DESTINATION ${CMAKE_INSTALL_PREFIX}
-            FILES_MATCHING PATTERN "*.h")
```
Comment (Baranowski) [inline cmake/Dependencies.cmake:859]: I didn't understand why pybind11 is installed so I put this change here as a placeholder but I'm sure there must be good reasons to keep installing pybind11. In that case, I can simply add some cached variable that remembers whether we should use `third_party/pybind11` or look for the library in the system.

This is important because when there is no pybind11 present in the system, on the first run `cmake` adds `-I third_party/pybind11/pybind11` to the compilation command, then installs pybind11 in the system (in case of `python setup.py develop` command, the destination is the `torch/` subfolder), and on subsequent runs changes the include path to that installation destination. This is probably wrong: if you update the `third_party/pybind11` submodule, incremental builds will not pick it up.

**Review by Baranowski — COMMENTED** [review]
@ezyang, There are three potential issues with this PR:

* There might be good reasons to keep installing pybind11 (details inline) (EDIT: this point is no longer an issue)
* There might be more bugs that manifest with different build environments (e.g., with different `USE_*` flags or on different OS's).
* Even after this change, there is some unnecessary work being done on the second (but not subsequent) build: 10 `.so` libraries are linked again. I left it that way because I think that's already a decent user experience, but I can dig into that as well.

Three of the snippets below are orthogonal to each other, and the fourth one depends on them. So I can split this PR.

**Comment by Baranowski** [comment]
Sorry, I see there is a lot of CI failures after I rebased and they look related. I will have to look at them first.

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (ezyang) [reply cmake/Dependencies.cmake:1133]: Eeeew. Can we just make it a hard error to enable HIP instead of toggling it OFF above?

**Comment by ezyang** [comment]
It seems this patch simplified a lot from earlier versions! Is it really the HIP toggling that is causing the rebuild?

**Comment by Baranowski** [comment]
Due to the CI failures, this is still a WiP. Sorry that I didn't make it clear before.

The patch looked simpler simpler simply because I'm trying to fix the CI failures, and so I'm selectively reverting some of the changes.

The HIP toggling itself causes rebuilds of about 800 files.

Change of pybind11 include directory and non-deterministic enabling of `THREADS_PREFER_PTHREAD_FLAG` cause a lot of rebuilds too.

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (Baranowski) [reply cmake/Dependencies.cmake:1138]: The CI depends on USE_ROCM being gracefully disabled. When I tried to hard error, there were 16 CI failures.

I could disable USE_ROCM by default and require the user to override it if desired. Would you like me to do that?

**Comment by Baranowski** [comment]
This is ready for a review

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (ezyang) [reply cmake/Dependencies.cmake:1138]: Let's take this as is for now. I see CUDA is doing a similar thing.

**Review by ezyang — APPROVED** [review]
